### PR TITLE
Rework login.lua and added other features

### PR DESCRIPTION
**Pull request by dex-89**
- [X] **Fixed login message**
- - [X] Welcome message added
- [X] **Premium account fixed**
- - [X] Premium Account ends: Teleport to Temple, change addon (citizen) houseless
- [X] **Recruiter system**
How it's working? You need to change the "recruiter" in the database...
The administrator decides what causes the update of this value
- [X] **Moved sendTibiaTime to cpp**
- [X] **Kick other players from account: only one player per account**
if set `onePlayerOnlinePerAccount` as `true`
- [X] **Block kick player if player hasn't PZ, kick the logger**

### CODE REVIEW

File: data/scripts/creaturescripts/others/login.lua [hunk 1]
```diff
@@ -23,26 +12,130 @@ local function onMovementRemoveProtection(cid, oldPos, time)
 
 	addEvent(onMovementRemoveProtection, 1000, cid, oldPos, time - 1)
 end
-
+local function antipz(p)
```
Comment (dudantas) [inline data/scripts/creaturescripts/others/login.lua]: I think we can rename this function to make it more readable.
Example:
antiProtectionZone
kickFromPz or kickFromProtectionZone

It's up to you

File: data/scripts/creaturescripts/others/login.lua [hunk 2]
```diff
@@ -23,26 +12,130 @@ local function onMovementRemoveProtection(cid, oldPos, time)
...
+end
 local playerLogin = CreatureEvent("PlayerLogin")
 function playerLogin.onLogin(player)
 	local items = {
 		{2120, 1},
 		{2148, 3}
 	}
+	local loginStr = "Welcome to " .. SERVER_NAME .. "!"	
 	if player:getLastLoginSaved() == 0 then
+		loginStr = loginStr .. " Please choose your outfit."
```
Comment (dudantas) [inline data/scripts/creaturescripts/others/login.lua]: This part of the login is incorrect. The server was adapted to start at dawnport and does not start with this option.

**Review by dudantas — CHANGES_REQUESTED** [review]
I made two observations, if you can analyze...

**Comment by dex-89** [comment]
> I made two observations, if you can analyze...

Fixed all 👍

**Comment by majestyotbr** [comment]
@dudantas PR waiting for your review.

File: data/scripts/creaturescripts/others/login.lua [hunk 3]
```diff
@@ -43,6 +37,101 @@ function playerLogin.onLogin(player)
...
+					addEvent(ProtectionZoneCheck, 2000, player:getName())
+					doPlayerPopupFYI(player, "You cant login now.")
+				end
+			until not result.next(resultId)
+				result.free(resultId)
+		end
+	end
+	-- end kick other players from account
+	if isPremium(player) then
+		player:setStorageValue(998899,1)
```
Comment (debone) [inline data/scripts/creaturescripts/others/login.lua]: What is this number `998899` ?

File: data/scripts/creaturescripts/others/login.lua [hunk 4]
```diff
@@ -43,6 +37,101 @@ function playerLogin.onLogin(player)
...
+        end
+        if home ~= nil and not isPremium(player) then
+            setHouseOwner(home, 0)
+            player:sendTextMessage(MESSAGE_STATUS_WARNING, 'You\'ve lost your house because you are not premium anymore.')
+			player:sendTextMessage(MESSAGE_STATUS_WARNING, 'Your items from house are send to your inbox.')
+        end
+	end
+	-- End 'Premium Ends Teleport to Temple'
+
+	-- Recruiter system
```
Comment (debone) [inline data/scripts/creaturescripts/others/login.lua:90]: What is "Recruiter system"?

File: src/protocolgame.cpp [hunk 5]
```diff
@@ -1655,6 +1655,14 @@ void ProtocolGame::sendWorldLight(const LightInfo& lightInfo)
 	writeToOutputBuffer(msg);
 }
 
+void ProtocolGame::sendTibiaTime(int32_t time) {
+	NetworkMessage msg;
+	msg.addByte(0xEF);
+	msg.addByte(time / 60);
+	msg.addByte(time % 60);
+	writeToOutputBuffer(msg);
```
Comment (debone) [inline src/protocolgame.cpp:1663]: What does this package do?

File: data/scripts/creaturescripts/others/login.lua [hunk 6]
```diff
@@ -43,6 +37,101 @@ function playerLogin.onLogin(player)
 	end
 
 	local playerId = player:getId()
+
+	-- kick other players from account
```
Comment (debone) [inline data/scripts/creaturescripts/others/login.lua:44]: Don't we block players from logging in if anyone on the account is online? That's how it works on global.

File: data/scripts/creaturescripts/others/login.lua [hunk 7]
```diff
@@ -43,6 +37,101 @@ function playerLogin.onLogin(player)
...
+		local Sex = player:getSex()
+		local home = getHouseByPlayerGUID(getPlayerGUID(player))
+		town = isInArray(freeTowns, town:getName()) and town or Town(defaultTown)
+		player:teleportTo(town:getTemplePosition())
+		player:setTown(town)
+		player:sendTextMessage(MESSAGE_STATUS_WARNING, "Your premium time has expired.")
+		player:setStorageValue(998899,0)
+		if Sex == 1 then
+			player:setOutfit({lookType = 128, lookFeet = 114, lookLegs = 134, lookHead = 114,lookAddons = 0})
+        end
```
Comment (debone) [inline data/scripts/creaturescripts/others/login.lua]: Tabs got messy from on here

File: data/scripts/creaturescripts/others/login.lua [hunk 8]
```diff
@@ -23,14 +12,19 @@ local function onMovementRemoveProtection(cid, oldPos, time)
 
 	addEvent(onMovementRemoveProtection, 1000, cid, oldPos, time - 1)
 end
-
+local function ProtectionZoneCheck(p)
```
Comment (debone) [inline data/scripts/creaturescripts/others/login.lua]: Maybe give this function another name? Why not call doRemoveCreature directly?

File: data/scripts/creaturescripts/others/login.lua [hunk 9]
```diff
@@ -43,6 +37,101 @@ function playerLogin.onLogin(player)
...
+		player:setTown(town)
+		player:sendTextMessage(MESSAGE_STATUS_WARNING, "Your premium time has expired.")
+		player:setStorageValue(998899,0)
+		if Sex == 1 then
+			player:setOutfit({lookType = 128, lookFeet = 114, lookLegs = 134, lookHead = 114,lookAddons = 0})
+        end
+        if Sex == 0 then
+			player:setOutfit({lookType = 136, lookFeet = 114, lookLegs = 134, lookHead = 114, lookAddons = 0})
+        end
+        if home ~= nil and not isPremium(player) then
```
Comment (debone) [inline data/scripts/creaturescripts/others/login.lua:82]: Is premium time deducted by players? I mean, if it depends on the player to login to kick them out of the house, what happens if they pay the rent, lose the premium and don't login?

File: data/scripts/creaturescripts/others/login.lua (hunk 3 above)
Comment (dex-89) [reply data/scripts/creaturescripts/others/login.lua]: his is the storage to check whether the player has a premium account or not. If not - moved to temple (other function: - Premium Ends Teleport to Temple, change addon (citizen) houseless)

File: data/scripts/creaturescripts/others/login.lua (hunk 4 above)
Comment (dex-89) [reply data/scripts/creaturescripts/others/login.lua:90]: https://tibia.fandom.com/wiki/Recruiter_Outfits

File: src/protocolgame.cpp (hunk 5 above)
Comment (dex-89) [reply src/protocolgame.cpp:1663]: https://github.com/opentibiabr/otservbr-global/blob/develop/data/scripts/creaturescripts/others/login.lua

```
function Player.sendTibiaTime(self, hours, minutes)
	-- TODO: Migrate to protocolgame.cpp
	local msg = NetworkMessage()
	msg:addByte(0xEF)
	msg:addByte(hours)
	msg:addByte(minutes)
	msg:sendToPlayer(self)
	msg:delete()
	return true
end
```
so i moved it to protocolgame.cpp

File: data/scripts/creaturescripts/others/login.lua (hunk 6 above)
Comment (dex-89) [reply data/scripts/creaturescripts/others/login.lua:44]: If you have another idea - let's go! :) That system works correctly but without it you can login with multiclient without errors

File: data/scripts/creaturescripts/others/login.lua (hunk 9 above)
Comment (dex-89) [reply data/scripts/creaturescripts/others/login.lua:82]: You have an example to do a check :)

**Review by dudantas — APPROVED** [review]

**Comment by dudantas** [comment]
> Fixed all 👍

Ok, thank for your contribution

File: src/protocolgame.cpp (hunk 5 above)
Comment (debone) [reply src/protocolgame.cpp:1663]: Sure, but now every light tick we send an update to every player which I'm not sure if it's required

Comment (debone) [reply src/protocolgame.cpp:1663]: And we cannot move things without understanding why and what for.

File: data/scripts/creaturescripts/others/login.lua (hunk 4 above)
Comment (debone) [reply data/scripts/creaturescripts/others/login.lua:90]: TIL

Haha my arm got tired just from seeing that outfit's arm holding the flag a meter from the body

File: data/scripts/creaturescripts/others/login.lua (hunk 3 above)
Comment (debone) [reply data/scripts/creaturescripts/others/login.lua]: Are there no "isPremium" functions? If surprisingly not, can we declare this number in some centralized file?

Comment (dex-89) [reply data/scripts/creaturescripts/others/login.lua]: That's corrected. this is moved to storage.lua

Comment (debone) [reply data/scripts/creaturescripts/others/login.lua]: Still, I think it's not necessary.

**Pull request merged**